File: src/lib/archive-timeline.ts
/**
 * Archiv-Timeline-Helfer: gruppiert die (client-seitig gelieferten) Story-Cards
 * nach Monat + Tag, berechnet den „Puls" (Kategorie-Verteilung je Tag/Monat) und
 * die Monats-Kennzahlen. Rein funktional, keine Seiteneffekte — von allen drei
 * Archiv-Konzepten (Puls / Spur / Logbuch) geteilt.
 */

export type ArchiveStory = {
	slug: string;
	title: string;
	dek: string;
	category: string;
	country: string;
	tone: 'amber' | 'sage' | 'rose' | 'sky';
	hero: string;
	impactScore: number;
	impactNote: string;
	sensitive?: boolean;
	publishedAt: string;
	createdAt?: string;
};

export type DayGroup = {
	iso: string; // YYYY-MM-DD (lokale Berlin-Zeit)
	date: Date;
	stories: ArchiveStory[];
	lead: ArchiveStory; // stärkste Story des Tages
	topCategory: string; // dominante Kategorie (für den Puls-Punkt)
};

export type MonthGroup = {
	key: string; // YYYY-MM
	label: string; // „Juni 2026"
	date: Date; // erster des Monats
	days: DayGroup[];
	count: number;
	topStory: ArchiveStory; // stärkste Story des Monats
	topImpact: number;
	categoryShare: { category: string; count: number; pct: number }[]; // Puls
	dominantCategory: string;
};

const MONTHS_DE = [
	'Januar', 'Februar', 'März', 'April', 'Mai', 'Juni',
	'Juli', 'August', 'September', 'Oktober', 'November', 'Dezember'
];

// Kategorie → Ton (für Puls-Farben; deckt sich mit story-card/toneStyles).
export const CATEGORY_TONE: Record<string, 'amber' | 'sage' | 'rose' | 'sky'> = {
	klima: 'sage',
	tiere: 'sage',
	gesundheit: 'rose',
	wissenschaft: 'sky',
	innovation: 'sky',
	gemeinschaft: 'amber',
	kultur: 'amber'
};

// Lokales (Europe/Berlin) YYYY-MM-DD — konsistent mit der Heute-Seite.
const BERLIN = new Intl.DateTimeFormat('en-CA', {
	timeZone: 'Europe/Berlin',
	year: 'numeric',
	month: '2-digit',
	day: '2-digit'
});
function berlinISO(d: Date): string {
	return BERLIN.format(d); // en-CA → „2026-07-09"
}

function impact(s: ArchiveStory): number {
	return s.impactScore ?? 0;
}

/** Dominante Kategorie einer Story-Liste (häufigste, bei Gleichstand stärkste). */
function dominant(stories: ArchiveStory[]): string {
	const count = new Map<string, number>();
	for (const s of stories) count.set(s.category, (count.get(s.category) ?? 0) + 1);
	let best = stories[0]?.category ?? 'gemeinschaft';
	let bestN = -1;
	for (const [cat, n] of count) {
		if (n > bestN) {
			best = cat;
			bestN = n;
		}
	}
	return best;
}

/**
 * Gruppiert Stories → Monate → Tage. Erwartet eine bereits gefilterte/sortierte
 * Liste ist NICHT nötig; wir sortieren selbst (neueste zuerst).
 */
export function groupByMonth(stories: ArchiveStory[]): MonthGroup[] {
	// Tag-Buckets
	const dayMap = new Map<string, ArchiveStory[]>();
	for (const s of stories) {
		const d = new Date(s.publishedAt || s.createdAt || Date.now());
		if (isNaN(d.getTime())) continue;
		const iso = berlinISO(d);
		const arr = dayMap.get(iso) ?? [];
		arr.push(s);
		dayMap.set(iso, arr);
	}

	// Tag-Gruppen bauen
	const days: DayGroup[] = [];
	for (const [iso, arr] of dayMap) {
		arr.sort((a, b) => impact(b) - impact(a));
		days.push({
			iso,
			date: new Date(iso + 'T12:00:00'),
			stories: arr,
			lead: arr[0],
			topCategory: dominant(arr)
		});
	}
	days.sort((a, b) => b.iso.localeCompare(a.iso)); // neueste zuerst

	// Monats-Buckets
	const monthMap = new Map<string, DayGroup[]>();
	for (const dg of days) {
		const key = dg.iso.slice(0, 7); // YYYY-MM
		const arr = monthMap.get(key) ?? [];
		arr.push(dg);
		monthMap.set(key, arr);
	}

	const months: MonthGroup[] = [];
	for (const [key, dayGroups] of monthMap) {
		const all = dayGroups.flatMap((d) => d.stories);
		const [y, m] = key.split('-').map(Number);
		const catCount = new Map<string, number>();
		for (const s of all) catCount.set(s.category, (catCount.get(s.category) ?? 0) + 1);
		const share = [...catCount.entries()]
			.map(([category, count]) => ({ category, count, pct: count / all.length }))
			.sort((a, b) => b.count - a.count);
		const top = all.slice().sort((a, b) => impact(b) - impact(a))[0];
		months.push({
			key,
			label: `${MONTHS_DE[m - 1]} ${y}`,
			date: new Date(y, m - 1, 1),
			days: dayGroups,
			count: all.length,
			topStory: top,
			topImpact: impact(top),
			categoryShare: share,
			dominantCategory: share[0]?.category ?? 'gemeinschaft'
		});
	}
	months.sort((a, b) => b.key.localeCompare(a.key)); // neueste zuerst
	return months;
}

/** Deutsche Wochentag-Kurzform + Tag-Zahl. */
export function dayParts(iso: string): { weekday: string; day: number; monthShort: string } {
	const d = new Date(iso + 'T12:00:00');
	const today = berlinISO(new Date());
	const yst = berlinISO(new Date(Date.now() - 864e5));
	let weekday: string;
	if (iso === today) weekday = 'Heute';
	else if (iso === yst) weekday = 'Gestern';
	else weekday = ['So', 'Mo', 'Di', 'Mi', 'Do', 'Fr', 'Sa'][d.getDay()];
	return {
		weekday,
		day: d.getDate(),
		monthShort: ['Jan', 'Feb', 'Mär', 'Apr', 'Mai', 'Jun', 'Jul', 'Aug', 'Sep', 'Okt', 'Nov', 'Dez'][d.getMonth()]
	};
}

File: ios-native/NurEine/NurEine/NurEine/Models/ArchiveTimeline.swift
import Foundation
import SwiftUI

/// Archiv-Timeline: gruppiert Stories nach Monat und Tag und berechnet den „Puls"
/// (Kategorie-Verteilung je Monat).
///
/// Swift-Gegenstück zu `src/lib/archive-timeline.ts` — die Website zeigt das
/// Archiv seit dem Redesign als Zeitreise (Monats-Kapitel statt Endlos-Grid),
/// und die App übernimmt genau diese Ordnung. Rein funktional, keine Seiteneffekte.

struct ArchiveDay: Identifiable {
    let iso: String            // YYYY-MM-DD (Berliner Zeit)
    let stories: [Story]       // stärkste zuerst
    let topCategory: String    // dominante Kategorie → Farbe des Tages-Punkts

    var id: String { iso }

    /// „Heute" / „Gestern" / „Mo" — die Website macht das genauso, es liest sich
    /// wärmer als ein reines Datum.
    var weekdayLabel: String {
        let today = ArchiveTimeline.berlinISO(Date())
        let yesterday = ArchiveTimeline.berlinISO(Date().addingTimeInterval(-86400))
        if iso == today { return "Heute" }
        if iso == yesterday { return "Gestern" }
        guard let d = ArchiveTimeline.date(fromISO: iso) else { return "" }
        let wd = Calendar(identifier: .gregorian).component(.weekday, from: d)
        return ["So", "Mo", "Di", "Mi", "Do", "Fr", "Sa"][max(0, min(6, wd - 1))]
    }

    /// „9. Jul"
    var dayLabel: String {
        guard let d = ArchiveTimeline.date(fromISO: iso) else { return "" }
        let cal = Calendar(identifier: .gregorian)
        let day = cal.component(.day, from: d)
        let month = cal.component(.month, from: d)
        let short = ["Jan", "Feb", "Mär", "Apr", "Mai", "Jun",
                     "Jul", "Aug", "Sep", "Okt", "Nov", "Dez"]
        return "\(day). \(short[max(0, min(11, month - 1))])"
    }
}

struct CategoryShare: Identifiable {
    let category: String
    let count: Int
    let pct: Double
    var id: String { category }
}

struct ArchiveMonth: Identifiable {
    let key: String                    // YYYY-MM
    let label: String                  // „Juli 2026"
    let days: [ArchiveDay]
    let count: Int
    let topStory: Story?               // Höhepunkt des Monats
    let categoryShare: [CategoryShare] // der „Puls"

    var id: String { key }
}

enum ArchiveTimeline {
    private static let monthsDE = [
        "Januar", "Februar", "März", "April", "Mai", "Juni",
        "Juli", "August", "September", "Oktober", "November", "Dezember"
    ]

    /// Kategorie → Markenfarbe. Deckt sich mit CATEGORY_TONE der Website.
    static func color(for category: String) -> Color {
        switch category {
        case "klima", "tiere": Theme.sage
        case "gesundheit": Theme.rose
        case "wissenschaft", "innovation": Theme.sky
        default: Theme.amber          // gemeinschaft, kultur
        }
    }

    private static let berlinFormatter: DateFormatter = {
        let f = DateFormatter()
        f.calendar = Calendar(identifier: .gregorian)
        f.locale = Locale(identifier: "en_US_POSIX")
        f.timeZone = TimeZone(identifier: "Europe/Berlin")
        f.dateFormat = "yyyy-MM-dd"
        return f
    }()

    static func berlinISO(_ d: Date) -> String { berlinFormatter.string(from: d) }
    static func date(fromISO iso: String) -> Date? { berlinFormatter.date(from: iso) }

    /// Die API liefert publishedAt als ISO-8601-String; wir brauchen nur den Tag.
    private static func dayKey(of story: Story) -> String? {
        let raw = story.publishedAt
        // Häufigster Fall: der String beginnt bereits mit YYYY-MM-DD.
        if raw.count >= 10 {
            let head = String(raw.prefix(10))
            if head.count == 10, head.dropFirst(4).first == "-" { return head }
        }
        return nil
    }

    /// Häufigste Kategorie einer Liste (bei Gleichstand die erste, also stärkste).
    private static func dominant(_ stories: [Story]) -> String {
        var counts: [String: Int] = [:]
        for s in stories { counts[s.category, default: 0] += 1 }
        return counts.max { a, b in a.value < b.value }?.key
            ?? stories.first?.category ?? "gemeinschaft"
    }

    /// Stories → Monate → Tage, jeweils neueste zuerst.
    static func groupByMonth(_ stories: [Story]) -> [ArchiveMonth] {
        // Tag-Eimer
        var dayMap: [String: [Story]] = [:]
        for s in stories {
            guard let key = dayKey(of: s) else { continue }
            dayMap[key, default: []].append(s)
        }

        let days: [ArchiveDay] = dayMap
            .map { iso, list in
                let sorted = list.sorted { $0.impactScore > $1.impactScore }
                return ArchiveDay(iso: iso, stories: sorted, topCategory: dominant(sorted))
            }
            .sorted { $0.iso > $1.iso }

        // Monats-Eimer
        var monthMap: [String: [ArchiveDay]] = [:]
        for d in days { monthMap[String(d.iso.prefix(7)), default: []].append(d) }

        return monthMap
            .map { key, dayGroups in
                let all = dayGroups.flatMap(\.stories)
                var counts: [String: Int] = [:]
                for s in all { counts[s.category, default: 0] += 1 }
                let share = counts
                    .map { CategoryShare(category: $0.key, count: $0.value,
                                         pct: Double($0.value) / Double(max(all.count, 1))) }
                    .sorted { $0.count > $1.count }

                let parts = key.split(separator: "-")
                let year = parts.first.map(String.init) ?? ""
                let monthIdx = (parts.count > 1 ? Int(parts[1]) : nil) ?? 1

                return ArchiveMonth(
                    key: key,
                    label: "\(monthsDE[max(0, min(11, monthIdx - 1))]) \(year)",
                    days: dayGroups,
                    count: all.count,
                    topStory: all.max { $0.impactScore < $1.impactScore },
                    categoryShare: share
                )
            }
            .sorted { $0.key > $1.key }
    }
}
